lib/Dialect/HW/HWAttributes.cpp

```diff
@@ -691,3 +691,79 @@ void ParamExprAttr::print(AsmPrinter &p) const {
+static Attribute
+replaceDeclRefInExpr(const std::map<std::string, IntegerAttr> &parameters,
+                     Attribute paramAttr) {
+  if (paramAttr.dyn_cast<IntegerAttr>()) {
+    // Nothing to do, constant value.
+    return paramAttr;
+  }
+  if (auto paramRefAttr = paramAttr.dyn_cast<hw::ParamDeclRefAttr>()) {
+    // Get the value from the provided parameters.
+    auto it = parameters.find(paramRefAttr.getName().str());
+    assert(it != parameters.end() && "Could not find parameter in the provided "
+                                     "parameters for the expression!");
+    return it->second;
+  }
+  if (auto paramExprAttr = paramAttr.dyn_cast<hw::ParamExprAttr>()) {
+    // Recurse into all operands of the expression.
+    llvm::SmallVector<Attribute, 4> replacedOperands;
+    llvm::transform(paramExprAttr.getOperands(),
+                    std::back_inserter(replacedOperands), [&](auto operand) {
+                      return replaceDeclRefInExpr(parameters, operand);
+                    });
+    return hw::ParamExprAttr::get(paramExprAttr.getOpcode(), replacedOperands);
+  }
+  assert(false && "Unhandled parametric attribute");
+  return {};
+}
+
+APInt hw::evaluateParamAttr(ArrayAttr parameters, Attribute paramAttr) {
+  // Create a map of the provided parameters for faster lookup.
+  std::map<std::string, IntegerAttr> parameterMap;
+  for (auto param : parameters) {
+    auto paramDecl = param.cast<ParamDeclAttr>();
+    auto paramValue = paramDecl.getValue().dyn_cast<IntegerAttr>();
+    assert(paramValue && "Expected parameter value to be a known integer");
+    parameterMap[paramDecl.getName().str()] = paramValue;
+  }
+
+  // First, replace any ParamDeclRefAttr in the expression with its
+  // corresponding value in 'parameters'.
+  paramAttr = replaceDeclRefInExpr(parameterMap, paramAttr);
+
+  // Then, evaluate the parametri attribute.
+  if (auto intAttr = paramAttr.dyn_cast<IntegerAttr>())
+    return intAttr.getValue();
+  if (auto paramExprAttr = paramAttr.dyn_cast<hw::ParamExprAttr>()) {
+    // Since any ParamDeclRefAttr was replaced within the expression, the
+    // expression should be able to be fully canonicalized to a constant. We do
+    // this through the existing ParamExprAttr canonicalizer.
+    auto resAttr = ParamExprAttr::get(paramExprAttr.getOpcode(),
+                                      paramExprAttr.getOperands());
+    auto resIntAttr = resAttr.dyn_cast<IntegerAttr>();
+    assert(resIntAttr &&
+           "Expected evaluated parametric expression to "
+           "canonicalize to a constant. Since not, this means that some parts "
+           "of the expression did not resolve to a constant");
+    return resIntAttr.getValue();
+  }
+  if (paramAttr.dyn_cast<hw::ParamDeclRefAttr>())
+    assert(false && "Should have been replaced earlier!");
+
+  assert(false && "Unhandled parametric attribute");
+  return APInt();
+}
+
+Type hw::resolveParametricType(ArrayAttr parameters, Type type) {
```

Comment:
Fully agree with your reasoning and to use FailureOr here. `resolveParametricType` and `evaluateParamAttr` should be able to be used liberally, so it doesn't really feel just to have these throw an assert in case of missing parameters.